torch/csrc/jit/script/init.cpp

```diff
@@ -329,20 +286,77 @@ struct ModuleValue : public SugaredValue {
     std::vector<std::shared_ptr<SugaredValue>> result;
     for(py::handle module : py_module) {
       py::object obj = py::reinterpret_borrow<py::object>(module);
-      if(py::isinstance<Module>(obj)) {
-        auto r = py::cast<std::shared_ptr<Module>>(obj);
-        result.push_back(std::make_shared<ModuleValue>(r));
-      } else {
-        result.push_back(ConstantPythonValue::create(loc, m, obj));
-      }
+      result.push_back(toSugaredValue(
+          obj,
+          m,
+          loc,
+          /*is_constant =*/false,
+          /*is_submodule =*/true));
     }
     return result;
   }
 
-private:
+ private:
   std::shared_ptr<Module> module;
 };
 
+std::shared_ptr<SugaredValue> toSugaredValue(
+    py::object obj,
+    Method& m,
+    SourceRange loc,
+    bool is_constant,
+    bool is_submodule) {
```

Comment:
This seems good to me. I see where `is_submodule` appeared. It is still only two context-sensitive flags, which is not too bad compared to what it was before.